src/data/data_store.cairo

```diff
@@ -549,8 +571,82 @@ mod DataStore {
             // Check that the caller has permission to set the value.
             self.role_store.read().assert_only_role(get_caller_address(), role::CONTROLLER);
 
-            // Set the value.
-            self.order_values.write(key, order);
+            let index_result = self.get_key_index(key);
+
+            match index_result {
+                // Update to new order for given key
+                Option::Some(i) => {
+                    self.order_values.write(key, order)
+                },
+                // Add new order for given key and increase the count
+                Option::None(()) => {
+                    let count = self.order_count.read();
+                    self.order_keys.write(count, key);
+                    self.order_count.write(count + 1);
+                    self.order_values.write(key, order);
+                },
+            };
+        }
+
+        fn remove_order(ref self: ContractState, key: felt252) {
+            // Check that the caller has permission to set the value.
+            self.role_store.read().assert_only_role(get_caller_address(), role::CONTROLLER);
+
+            let order = self.order_values.read(key);
+            // We use the zero address to indicate that the order does not exist.
+            assert(!order.account.is_zero(), DataError::ORDER_NOT_FOUND);
+
+            let index_result = self.get_key_index(key);
+            let mut index: u64 = 0;
+            match index_result {
+                Option::Some(i) => index = i,
+                Option::None(()) => assert(false, DataError::ORDER_INDEX_NOT_FOUND),
+            };
+
+            let count = self.order_count.read();
+            let last_key = self.order_keys.read(count - 1);
+
+            // Update removed key index with last key
```

Comment:
Is there a requirement to preserve Orders order?

Reply:
no. I am moving key from last index to removed key index. Do you have better  idea?

Reply:
No I think it's a good idea